File: tree/phone_number_list.cpp
/*
TITLE   : 전화번호 목록
URL     : https://www.acmicpc.net/problem/5052
DATE    : 21.07.14
*/
#include <iostream>
#include <string>
#include <string.h>

using namespace std;

struct Node{
    bool is_end;
    Node *next[10];
    Node():is_end(false){
        memset(next,0,sizeof(next));
    }
    ~Node(){
        for(int i=0; i < 10; ++i){
			if(next[i])
				delete next[i];
		}
    }
};

class Dictree{
private:
    Node* root;
public:
    Dictree(){
        root = new Node;
    }
    bool insert(Node *curr,char* num){
        cout << curr->is_end;
        if(curr->is_end)
            return false;
        if(*num == '\n'){
            curr->is_end = true;
            return true;
        }
        else{
            int curr_num = *num - '0';
            if(curr->next[curr_num]==nullptr)
                curr->next[curr_num] = new Node;
            return insert(curr->next[curr_num], num + 1);
        }
    }
    Node* get_root(){
        return root;
    }
};

int main(){
    int t=0;
    scanf("%d",&t);
    while(t--){
        int n=0;
        scanf("%d",&n);
        Dictree tree;
        for(int i=0; i<n; i++){
            char* s;
            scanf("%s",s);
            bool valid = tree.insert(tree.get_root(),s);
            if(valid)
                printf("valid!\n");
            else
                printf("not valid!\n");
        }
    }

    return 0;
}

File: tree/phone_number_list.py
'''
TITLE   : 전화번호 목록
URL     : https://www.acmicpc.net/problem/5052
DATE    : 21.07.15
'''
import sys

t = int(sys.stdin.readline())

class Node(object):
    def __init__(self, key, data=None):
        self.key = key
        self.data = data
        self.next = {}

class Trie(object):
    def __init__(self):
        self.root = Node(None)
    def insert(self, numbers):
        curr = self.root
        for number in numbers:
            if number not in curr.next:
                curr.next[number] = Node(number)
            curr = curr.next[number]
        curr.data = numbers
    def is_valid(self, number):
        curr = self.root
        leng = len(number)
        for i in range(leng):
            if number[i] in curr.next:
                curr = curr.next[number[i]]
                if curr.data != None and i<leng-1:
                    return False
            else:
                return print("number is not in")
        return True

for _ in range(t):
    n = int(sys.stdin.readline())
    numbers = []
    for i in range(n):
        numbers.append(sys.stdin.readline().strip())
    number_list = Trie()
    for i in range(n):
        number_list.insert(numbers[i])
    is_valid = True
    for i in range(n):
        if not number_list.is_valid(numbers[i]):
            is_valid = False
            break
    if is_valid:
        print("YES")
    else:
        print("NO")